File: perf_test/perf_test.py
#!/usr/bin/env python

import subprocess

from datetime import datetime as dt


class PerfTest:
    def run(self):
        test_time = dt.now().strftime('%Y-%m-%d_%H-%M-%S.%f')
        fpath = f'/tmp/PerfTest_{test_time}.csv'

        build_type = 'release'
        compiler = 'g++-12'
        iters = 10
        always_inline = ""
        inline = ""

        versions = ['v06', 'v07', 'v08', 'v09', 'v10',
                    'v11', 'v12', 'v13', 'v14', 'v15',
                    'v16', 'v17', 'v18', 'v19', 'v20',
                    'v21', 'v22', 'v23', 'v24', 'v25']

        self._execute('rm -fr build')

        for version in versions[:3]:
            if version < 'v12':
                meta_suffix = ''
                bin_suffix = '-sorted'
                order_add = 'ItchOrderAdd'
                order_replace = 'ItchOrderReplace'
            elif version < 'v17':
                meta_suffix = ''
                bin_suffix = '-sorted-idx'
                order_add = 'ItchOrderAddIdx'
                order_replace = 'ItchOrderReplaceIdx'
            else:
                meta_suffix = '-reverse-bid'
                bin_suffix = '-sorted-idx-reverse-bid'
                order_add = 'ItchOrderAddIdx'
                order_replace = 'ItchOrderReplaceIdx'

            self._update_file(version, order_add, order_replace, always_inline, inline)
            self._rebuild(build_type, compiler)
            print(fpath)
            self._run_test(fpath, compiler, version, meta_suffix, bin_suffix, iters)
        print(fpath)

    def _run_test(self, fpath, test_id, version, meta_suffix, bin_suffix, iters):
        cmd = f'build/perf_test/perf_test {fpath} {test_id} {version} "{meta_suffix}" "{bin_suffix}" {iters}'
        self._execute(cmd)

    def _rebuild(self, build_type, compiler):
        cmd = f'cmake -S .. -B build -DCMAKE_BUILD_TYPE="{build_type}" -DCMAKE_CXX_COMPILER="{compiler}"'
        self._execute(cmd)
        cmd = f'cmake --build build --target perf_test'
        self._execute(cmd)

    def _update_file(self, version, order_add, order_replace, always_inline, inline):
        self._execute("cp -f perf_test.cxx perf_test.cpp")
        self._sed(f"'s/v11/{version}/g'")
        self._sed(f"'s/ItchOrderAdd/{order_add}/g'")
        self._sed(f"'s/ItchOrderReplace/{order_replace}/g'")
        self._sed(f"'s/__REPLACE__ALWAYS_INLINE__/{always_inline}/g'")
        self._sed(f"'s/__REPLACE__INLINE__/{inline}/g'")

    def _sed(self, sed):
        self._execute(f"sed -i {sed} perf_test.cpp")

    def _execute(self, cmd):
        subprocess.check_call(cmd, shell=True)


if __name__ == "__main__":
    PerfTest().run()

File: perf_test/perf_test.cxx
#define ALWAYS_INLINE __REPLACE__ALWAYS_INLINE__
#define INLINE __REPLACE__INLINE__

#include "itch/common.hpp"
#include "itch/itch.hpp"
#include "itch/metadata_io.hpp"

#include "itch/v26/order_book.hpp"

#include "nostromo/mmap.hpp"
#include "nostromo/time_utils.hpp"
#include "nostromo/huge_page.hpp"
#include "nostromo/thread_utils.hpp"

#include <fmt/format.h>

#include <cstdint>
#include <stdexcept>
#include <string>

namespace order_book::itch::perf_test {

struct Args {
   std::string fpath = "/tmp/PerfTest-default.csv";
   std::string id = "default";
   std::string version = "v26";
   std::string meta_suffix = "-reverse-bid";
   std::string bin_suffix = "-sorted-idx-reverse-bid";
   int iters = 1;
};

class PerfTest {
   const Args &args_;
   const std::string &fname_;
   std::ofstream &out_;

public:
   explicit PerfTest(
         const Args &args,
         const std::string &fname,
         std::ofstream &out)
         : args_{args},
           fname_{fname},
           out_{out} {}

   template<typename BOOKS>
   void Execute(
         const size_t orders_page_size = nos::HugePage::SIZE_1GB,
         const size_t other_page_size = nos::HugePage::SIZE_2MB) {
      const auto start = nos::TimeUtils::Now();
      const auto fprefix = itch::DATA_DIR_BASE + fname_;

      const auto [
            max_order_id,
            max_stock_code,
            stock_price_map
      ] = MetadataIO::Read(fprefix + ".meta" + args_.meta_suffix);

      BOOKS books{max_order_id, max_stock_code, stock_price_map, orders_page_size, other_page_size};

      const nos::Mmap<char> mmap{fprefix + ".bin" + args_.bin_suffix};
      const auto data = mmap.Span();

      uint64_t offset = 0;
      uint64_t order_cnt = 0;

      while (offset < data.size()) {
         const auto msg_type = data[offset];
         ++offset;
         ++order_cnt;

         switch (msg_type) {
            case 'A':
            case 'F': {
               const auto order = reinterpret_cast<ItchOrderAdd *>(&data[offset]);
               books.OrderAdd(*order);
               offset += sizeof(ItchOrderAdd);
               break;
            }
            case 'E':
            case 'C': {
               const auto order = reinterpret_cast<ItchOrderExecuted *>(&data[offset]);
               books.OrderExecuted(*order);
               offset += sizeof(ItchOrderExecuted);
               break;
            }
            case 'X': {
               const auto order = reinterpret_cast<ItchOrderCancel *>(&data[offset]);
               books.OrderCancel(*order);
               offset += sizeof(ItchOrderCancel);
               break;
            }
            case 'D': {
               const auto order = reinterpret_cast<ItchOrderDelete *>(&data[offset]);
               books.OrderDelete(*order);
               offset += sizeof(ItchOrderDelete);
               break;
            }
            case 'U': {
               const auto order = reinterpret_cast<ItchOrderReplace *>(&data[offset]);
               books.OrderReplace(*order);
               offset += sizeof(ItchOrderReplace);
               break;
            }
            default:
               throw std::runtime_error("unsupported msg_type at offset: " + std::to_string(offset - 1));
         }
      }

      const auto elap = nos::TimeUtils::Now() - start;
      const auto npo = static_cast<double>(elap.count()) / static_cast<double>(order_cnt);

      out_ << args_.id << "," << args_.version << "," << fname_ << "," <<
           elap.count() << "," << order_cnt << "," << npo << std::endl;
      fmt::print("version: {}  ns/order: {:.4f}\n", args_.version, npo);
   }

   static auto ParseArgs(int argc, char **argv) {
      Args args{};

      if (argc == 7) {
         args.fpath = argv[1];
         args.id = argv[2];
         args.version = argv[3];
         args.meta_suffix = argv[4];
         args.bin_suffix = argv[5];
         args.iters = std::stoi(argv[6]);
      }

      return args;
   }
};

} // namespace order_book::itch::perf_test

int main(int argc, char **argv) {
   namespace itch = order_book::itch;

   const auto args = itch::perf_test::PerfTest::ParseArgs(argc, argv);
   std::ofstream out(args.fpath, std::ios_base::app);

   const auto cpuid = static_cast<int>(std::thread::hardware_concurrency()) - 1;
   fmt::print("using cpuid: {}\n", cpuid);
   nos::ThreadUtils::SetAffinity(cpuid);

   for (const auto &fname: itch::DATA_FILE_NAMES) {
      fmt::print("processing: {}\n", fname);
      auto test = itch::perf_test::PerfTest{args, fname, out};

      for (auto x = 0; x < args.iters; ++x) {
         test.Execute<itch::v26::OrderBooks>();
      }
   }
}
